libs/content/workspace/src/task_manager.rs

```diff
@@ -0,0 +1,462 @@
+use std::collections::HashMap;
+use std::sync::{mpsc, Arc, Mutex};
+use std::time::Instant;
+use std::{mem, thread};
+
+use egui::Context;
+use lb_rs::blocking::Lb;
+use lb_rs::logic::crypto::DecryptedDocument;
+use lb_rs::model::errors::LbResult;
+use lb_rs::model::file_metadata::DocumentHmac;
+use lb_rs::service::sync::{SyncProgress, SyncStatus};
+use lb_rs::Uuid;
+
+pub struct TaskManager {
+    // queue tasks then call `update` to launch them
+    queued_loads: Vec<QueuedLoad>,
+    queued_saves: Vec<QueuedSave>,
+    queued_syncs: Vec<QueuedSync>,
+
+    // launched tasks stored here until complete
+    pub in_progress_loads: Vec<InProgressLoad>,
+    pub in_progress_saves: Vec<InProgressSave>,
+    pub in_progress_sync: Option<InProgressSync>,
+
+    // completions stored here then returned in the response on the next frame
+    completed_loads: Vec<CompletedLoad>,
+    completed_saves: Vec<CompletedSave>,
+    completed_sync: Option<CompletedSync>,
+}
+
+struct InnerResponse {
+    loads_to_launch: Vec<QueuedLoad>,
```

Comment:
`loads_to_launch` 😏 

it's not clear to me why loading can't begin immediately making progress. Having to wait for a frame to do this (and has no implication for the UI) seems like a code smell to me.

In my head I'm thinking about something that looks mostly like this but operates in a more channel based manner. And you don't need the UI thread as this sort of touch point for progress to happen / state communication to be facilitated.

Would be mostly the same API (with many of the same wins of centrally managing this complexity here) but UI would be able to receive at it's convenience, and threads would make progress optimally.

Reply:
Okay I'll have another pass with attention to the involvement in the UI thread in making progress